apigw/src/shared/auth/keycloak-saml.ts

```diff
@@ -0,0 +1,63 @@
+import {
+  Profile,
+  Strategy as SamlStrategy,
+  VerifiedCallback
+} from 'passport-saml'
+import { SamlUser } from '../routes/auth/saml/types'
+import { getOrCreateEmployee } from '../service-client'
+
+const host = 'http://localhost:9093'
+const keycloackSamlCallbackUrl = `${host}/api/internal/auth/evaka/login/callback`
+const keycloakSamlIssuer = 'evaka'
+
+// The magic url
+const keycloakEntrypoint =
+  'http://localhost:8080/auth/realms/evaka/protocol/saml'
+
+export default function createKeycloakSamlStrategy(): SamlStrategy {
+  return new SamlStrategy(
+    {
+      issuer: keycloakSamlIssuer,
+      callbackUrl: keycloackSamlCallbackUrl,
+      entryPoint: keycloakEntrypoint,
+      logoutUrl: keycloakEntrypoint,
+      acceptedClockSkewMs: -1
+    },
+    (profile: Profile, done: VerifiedCallback) => {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      verifyKeycloakProfile(profile as KeycloakProfile)
+        .then((user) => done(null, user))
+        .catch(done)
+    }
+  )
+}
+
+interface KeycloakProfile {
+  nameID: string
+  ID: string
+  nameIDFormat?: string
+  nameQualifier?: string
+  spNameQualifier?: string
+  sessionIndex?: string
+}
+
+async function verifyKeycloakProfile(
+  profile: KeycloakProfile
+): Promise<SamlUser> {
+  const person = await getOrCreateEmployee({
+    externalId: `evaka:${profile.ID}`,
```

Comment:
I strongly recommend adding a null/empty check for `ID` (and maybe even making it `ID?: string` in the `KeycloakProfile` type). Otherwise, any kind of configuration error or bug might lead to everybody logging in with the same "bugged" external id `evaka:undefined` or `evaka:null` or `evaka:`, which are all valid IDs and will point to the same employee(s).